### PR TITLE
migration to new design transformer

### PR DESCRIPTION
**Pull request by anistouri**
Signed-off-by: TOURI ANIS <anis-1.touri@rte-france.com>

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java [hunk 1]
```diff
@@ -107,151 +107,58 @@ public class TwoWindingsTransformerCreationEntity extends BranchCreationEntity {
...
+        TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos = (TwoWindingsTransformerCreationInfos) modificationInfos;
+        assignTapChanger(twoWindingsTransformerCreationInfos);
     }
 
-    @Override
-    public TwoWindingsTransformerCreationInfos toModificationInfos() {
-        return toTwoWindingsTransformerCreationInfosBuilder().build();
+    public TwoWindingsTransformerCreationEntity(TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos) {
+        super(twoWindingsTransformerCreationInfos);
+        this.magnetizingConductance = twoWindingsTransformerCreationInfos.getMagnetizingConductance();
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:119]: Add a private method `assignAttrributes`

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java [hunk 2]
```diff
@@ -107,151 +107,58 @@ public class TwoWindingsTransformerCreationEntity extends BranchCreationEntity {
...
-        this.ratioTapChangerTerminalRefVoltageLevelId = ratioTapChangerTerminalRefVoltageLevelId;
-        this.ratioTapChangerTerminalRefType = ratioTapChangerTerminalRefType;
-        this.ratioTapChangerLoadTapChangingCapabilities = ratioTapChangerLoadTapChangingCapabilities;
-        this.ratioTapChangerTargetV = ratioTapChangerTargetV;
-        this.tapChangerSteps = tapChangerSteps;
+    @Override
+    public void update(ModificationInfos modificationInfos) {
+        super.update(modificationInfos);
+        TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos = (TwoWindingsTransformerCreationInfos) modificationInfos;
+        assignTapChanger(twoWindingsTransformerCreationInfos);
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:114]: call `assignAttrributes` for other attributes

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java [hunk 3]
```diff
@@ -107,151 +107,58 @@ public class TwoWindingsTransformerCreationEntity extends BranchCreationEntity {
...
-                ratioTapChangerInfos != null ? ratioTapChangerInfos.getRegulatingTerminalVlId() : null,
-                ratioTapChangerInfos != null ? ratioTapChangerInfos.getRegulatingTerminalType() : null,
-                ratioTapChangerInfos != null ? ratioTapChangerInfos.isLoadTapChangingCapabilities() : null,
-                ratioTapChangerInfos != null ? ratioTapChangerInfos.getTargetV() : null,
-                tapChangerSteps,
-                twoWindingsTransformerCreationInfos.getConnectionPosition1(),
-                twoWindingsTransformerCreationInfos.getConnectionPosition2());
-
-        return twoWindingsTransformerCreationEntity;
+    private void assignRatioTapChanger(RatioTapChangerCreationInfos ratioTapChanger) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:133]: Rename by `ratioTapChangerInfos` ?

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java [hunk 4]
```diff
@@ -107,151 +107,58 @@ public class TwoWindingsTransformerCreationEntity extends BranchCreationEntity {
...
+        this.ratioTapChangerTargetDeadband = ratioTapChanger.getTargetDeadband();
+        this.ratioTapChangerTerminalRefConnectableId = ratioTapChanger.getRegulatingTerminalId();
+        this.ratioTapChangerTerminalRefVoltageLevelId = ratioTapChanger.getRegulatingTerminalVlId();
+        this.ratioTapChangerTerminalRefType =  ratioTapChanger.getRegulatingTerminalType();
+        this.ratioTapChangerLoadTapChangingCapabilities = ratioTapChanger.isLoadTapChangingCapabilities();
+        this.ratioTapChangerTargetV =  ratioTapChanger.getTargetV();
+        this.tapChangerSteps.addAll(TapChangerStepCreationEmbeddable.toEmbeddableRatioTapChangerSteps(ratioTapChanger.getSteps()));
+    }
+
+    private void assignPhaseTapChanger(PhaseTapChangerCreationInfos phaseTapChangerCreationInfos) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:146]: Rename by `phaseTapChangerInfos` ?

File: src/main/java/org/gridsuite/modification/server/modifications/EquipmentAttributeModification.java [hunk 5]
```diff
@@ -15,7 +15,8 @@
 import org.gridsuite.modification.server.NetworkModificationException;
 import org.gridsuite.modification.server.dto.EquipmentAttributeModificationInfos;
 
-import static org.gridsuite.modification.server.NetworkModificationException.Type.*;
+import static org.gridsuite.modification.server.NetworkModificationException.Type.EQUIPMENT_NOT_FOUND;
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/EquipmentAttributeModification.java:18]: Do not modify only imports in a class

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+        }
+
+        // BranchAdder completion by topology
+        setBranchAdderNodeOrBus(branchAdder, voltageLevel1, twoWindingsTransformerCreationInfos, Branch.Side.ONE, withSwitch1);
+        setBranchAdderNodeOrBus(branchAdder, voltageLevel2, twoWindingsTransformerCreationInfos, Branch.Side.TWO, withSwitch2);
+
+        return twoWindingsTransformerAdder;
+    }
+
+    private void setBranchAdderNodeOrBus(BranchAdder<?> branchAdder, VoltageLevel voltageLevel, BranchCreationInfos branchCreationInfos, Branch.Side side, boolean withSwitch) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java:92]: Move it to `ModificationUtils` class

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+        if (voltageLevel.getTopologyKind() == TopologyKind.NODE_BREAKER) {
+            setBranchAdderNode(branchAdder, voltageLevel, branchCreationInfos, currentBusBarSectionId, side, withSwitch);
+        } else { // BUS BREAKER
+            setBranchAdderBus(branchAdder, voltageLevel, currentBusBarSectionId, side);
+        }
+
+    }
+
+    private void setBranchAdderBus(BranchAdder<?> branchAdder, VoltageLevel voltageLevel, String currentBusBarSectionId, Branch.Side side) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java:103]: Move it to `ModificationUtils` class

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+        // complete the lineAdder
+        if (side == Branch.Side.ONE) {
+            branchAdder.setBus1(bus.getId()).setConnectableBus1(bus.getId());
+        } else {
+            branchAdder.setBus2(bus.getId()).setConnectableBus2(bus.getId());
+        }
+    }
+
+    private void setBranchAdderNode(BranchAdder<?> branchAdder, VoltageLevel voltageLevel, BranchCreationInfos branchCreationInfos, String currentBusBarSectionId, Branch.Side side, boolean withSwitch) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java:115]: Move it to `ModificationUtils` class

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        this.modificationInfos = modificationInfos;
+    }
+
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        // create the 2wt in the network
+        VoltageLevel voltageLevel1 = ModificationUtils.getInstance().getVoltageLevel(network, modificationInfos.getVoltageLevelId1());
+        VoltageLevel voltageLevel2 = ModificationUtils.getInstance().getVoltageLevel(network, modificationInfos.getVoltageLevelId2());
+        if (voltageLevel1.getTopologyKind() == TopologyKind.NODE_BREAKER && voltageLevel2.getTopologyKind() == TopologyKind.NODE_BREAKER) {
+            var twoWindingsTransformerAdder = createTwoWindingsTransformerAdder(network, voltageLevel1, voltageLevel2, modificationInfos, false, false);
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java:30]: Create a private method `create2WTInBusBreaker`

File: src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java [hunk 10]
```diff
@@ -993,145 +993,6 @@ public void testCreateGeneratorInBusBreaker() throws Exception {
...
-
-        testNetworkModificationsCount(TEST_GROUP_ID, 1);
-        mvcResult = mockMvc.perform(get(URI_NETWORK_MODIF_GET_PUT + bsmlrTwoWindings.get(0).getUuid()))
-                .andExpect(status().isOk()).andReturn();
-        resultAsString = mvcResult.getResponse().getContentAsString();
-        TwoWindingsTransformerCreationInfos bsmlrWindingsTransformer = mapper.readValue(resultAsString, new TypeReference<>() { });
-        assertThat(bsmlrWindingsTransformer, MatcherTwoWindingsTransformerCreationInfos.createMatcherTwoWindingsTransformerCreationInfos(twoWindingsTransformerCreationInfos));
-
-        // create 2wt with errors
-        twoWindingsTransformerCreationInfos.setBusOrBusbarSectionId1("notFoundBus");
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java:1127]: Where is this test ?

File: src/test/java/org/gridsuite/modification/server/BuildTest.java [hunk 11]
```diff
@@ -449,14 +550,6 @@ public void runBuildTest() throws Exception {
         assertEquals("v1", network.getLine("newLine").getTerminal1().getVoltageLevel().getId());
         assertEquals("v2", network.getLine("newLine").getTerminal2().getVoltageLevel().getId());
         assertEquals(4., network.getLine("newLine").getB1(), 0.1);
-        assertEquals("v1", network.getTwoWindingsTransformer("new2wt").getTerminal1().getVoltageLevel().getId());
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/BuildTest.java:452]: Why do you remove these tests
Move them in the `TwoWindingsTransformerCreationTest` class

File: src/test/java/org/gridsuite/modification/server/ModificationRepositoryTest.java [hunk 12]
```diff
@@ -412,56 +412,6 @@ public void testLineCreation() {
         );
     }
 
-    @Test
-    public void testTwoWindingsTransformerCreation() {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/ModificationRepositoryTest.java:416]: Why do remove this test ?

File: src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java [hunk 13]
```diff
@@ -1208,7 +1069,7 @@ private List<TapChangerStepCreationInfos> getTapChangerSteps() {
         );
     }
 
-    private void testCreateTwoWindingsTransformerInNodeBreaker(TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos) throws Exception {
+    private void testCreateTwoWindingsTransformerInNodeBreaker(TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos, int actualSize) throws Exception {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java:1072]: Move all tests `testCreateTwoWindingsTransformer*` in TwoWindingsTransformerCreationTest class

**Review by SlimaneAmar — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 14]
```diff
@@ -0,0 +1,265 @@
+package org.gridsuite.modification.server.modifications;
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:1]: Add Copyright

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java [hunk 15]
```diff
@@ -0,0 +1,217 @@
+package org.gridsuite.modification.server.modifications;
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreation.java:1]: add copyright header for new file

File: src/test/java/org/gridsuite/modification/server/BuildTest.java [hunk 16]
```diff
@@ -364,7 +364,108 @@ public void runBuildTest() throws Exception {
...
+                .currentLimits2(CurrentLimitsInfos.builder().permanentLimit(2.).build())
+                .connectionName1("cn201")
+                .connectionDirection1(ConnectablePosition.Direction.TOP)
+                .connectionName2("cn202")
+                .connectionDirection2(ConnectablePosition.Direction.TOP)
+                .phaseTapChanger(PhaseTapChangerCreationInfos.builder()
+                        .lowTapPosition(1)
+                        .tapPosition(2)
+                        .regulatingTerminalId("v1load")
+                        .regulatingTerminalVlId("v1")    .regulating(false)
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/BuildTest.java:392]: .regulating(false) on next line

File: src/test/java/org/gridsuite/modification/server/BuildTest.java [hunk 17]
```diff
@@ -364,7 +364,108 @@ public void runBuildTest() throws Exception {
...
+                        .regulatingTerminalType("LOAD")  .regulationMode(PhaseTapChanger.RegulationMode.CURRENT_LIMITER)
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/BuildTest.java:393]: .regulationMode(...) on next line

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+                .currentLimits2(CurrentLimitsInfos.builder().permanentLimit(2.).build())
+                .connectionName1("cn201")
+                .connectionDirection1(ConnectablePosition.Direction.TOP)
+                .connectionName2("cn202")
+                .connectionDirection2(ConnectablePosition.Direction.TOP)
+                .phaseTapChanger(PhaseTapChangerCreationInfos.builder()
+                        .lowTapPosition(1)
+                        .tapPosition(2)
+                        .regulatingTerminalId("v1load")
+                        .regulatingTerminalVlId("v1")    .regulating(false)
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:57]: put .regulating(false) in new line

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+                        .regulatingTerminalType("LOAD")  .regulationMode(PhaseTapChanger.RegulationMode.CURRENT_LIMITER)
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:58]: put .regulationMode( in new line

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+                                        .build()
+                        ))
+                        .build())
+                .build();
+    }
+
+    @Override
+    protected ModificationInfos buildModificationUpdate() {
+        return TwoWindingsTransformerCreationInfos.builder()
+                .type(ModificationType.TWO_WINDINGS_TRANSFORMER_CREATION)
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:137]: this update-case DTO is an exact copy/paste from the creation-case DTO.
you must change some values

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+    @Test
+    public void testCreateWithErrors() {
+        TwoWindingsTransformerCreationInfos twoWindingsTransformerCreationInfos = (TwoWindingsTransformerCreationInfos) buildModification();
+        twoWindingsTransformerCreationInfos.setEquipmentId("");
+        String substationCreationInfosJson = mapper.writeValueAsString(twoWindingsTransformerCreationInfos);
+        MvcResult mvcResult = mockMvc.perform(post(getNetworkModificationUri()).content(substationCreationInfosJson).contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().is5xxServerError()).andReturn();
+        String resultAsString = mvcResult.getResponse().getContentAsString();
+        assertEquals(resultAsString, new NetworkModificationException(NetworkModificationException.Type.CREATE_TWO_WINDINGS_TRANSFORMER_ERROR, "Invalid id ''").getMessage());
+    }
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:263]: in previous unit test, there is an error test case with BUS_NOT_FOUND.
can we have it here too ?

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+
+    @Override
+    protected MatcherTwoWindingsTransformerCreationInfos createMatcher(ModificationInfos modificationInfos) {
+        return MatcherTwoWindingsTransformerCreationInfos.createMatcherTwoWindingsTransformerCreationInfos((TwoWindingsTransformerCreationInfos) modificationInfos);
+    }
+
+    @Override
+    protected void assertNetworkAfterCreation() {
+        assertNotNull(getNetwork().getTwoWindingsTransformer("new2wt"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:245]: modification-server traces: "2 voltage level resources updated in 50 ms"
We can assert theses changes here:
```
assertEquals(1, getNetwork().getVoltageLevel("v1").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).cou
assertEquals(1, getNetwork().getVoltageLevel("v2").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).count());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 23]
```diff
@@ -0,0 +1,265 @@
...
+    }
+
+    @Override
+    protected void assertNetworkAfterDeletion() {
+        assertNull(getNetwork().getTwoWindingsTransformer("new2wt"));
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:250]:
```
assertEquals(0, getNetwork().getVoltageLevel("v1").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).count());
assertEquals(0, getNetwork().getVoltageLevel("v2").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).count());
```

**Review by dbraquart — CHANGES_REQUESTED** [review]
tests OK
review: just a few changes to address

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java (hunk 3 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:133]: We are mapping the data to the entity, we can keep it like this. are you ok ?

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java (hunk 4 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:146]: same answer

File: src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java (hunk 3 above)
Comment (dbraquart) [reply src/main/java/org/gridsuite/modification/server/entities/equipment/creation/TwoWindingsTransformerCreationEntity.java:137]: to be consistent with assignPhaseTapChanger:
ratioTapChanger -> ratioTapChangerCreationInfos

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java (hunk 20 above)
Comment (dbraquart) [reply src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:151]: you should change even more data, as mush as you can

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 24]
```diff
@@ -0,0 +1,558 @@
...
+        assertEquals(5., getNetwork().getTwoWindingsTransformer("new2wt").getRatedU1(), 0.1);
+        assertEquals(1, getNetwork().getTwoWindingsTransformer("new2wt").getRatedS(), 0.1);
+        assertEquals(4, getNetwork().getTwoWindingsTransformer("new2wt").getRatioTapChanger().getStepCount());
+        assertEquals(3, getNetwork().getTwoWindingsTransformer("new2wt").getPhaseTapChanger().getStepCount());
+        assertEquals(PhaseTapChanger.RegulationMode.CURRENT_LIMITER, getNetwork().getTwoWindingsTransformer("new2wt").getPhaseTapChanger().getRegulationMode());
+    }
+
+    @SneakyThrows
+    @Override
+    protected void assertNetworkAfterDeletion() {
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:275]: you can also put back :
`assertNull(getNetwork().getTwoWindingsTransformer("new2wt"));`

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java [hunk 25]
```diff
@@ -0,0 +1,558 @@
...
+        assertEquals(0, getNetwork().getVoltageLevel("v1").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).count());
+        assertEquals(0, getNetwork().getVoltageLevel("v2").getTwoWindingsTransformerStream().filter(transformer -> transformer.getId().equals("new2wt")).count());
+        testNetworkModificationsCount(getGroupId(), 0);
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationTest.java:278]: testNetworkModificationsCount is not a network assertion: remove it

**Review by dbraquart — CHANGES_REQUESTED** [review]
UI tests: todo
review:
- an issue related to assignAttributes, and the update test case
- all xxxInBusBreaker tests are missing, and you should have 2 files like other modifications: TwoWindingsTransformerCreationInNodeBreakerTest + TwoWindingsTransformerCreationInBusBreakerTest

File: src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java [hunk 26]
```diff
@@ -993,351 +993,6 @@ public void testCreateGeneratorInBusBreaker() throws Exception {
...
-                .connectionDirection1(ConnectablePosition.Direction.TOP)
-                .connectionName2("cnid2wt2")
-                .connectionDirection2(ConnectablePosition.Direction.TOP)
-                .phaseTapChanger(phaseTapChangerCreationInfos)
-                .build();
-        testCreateTwoWindingsTransformerInNodeBreaker(twoWindingsTransformerCreationInfos);
-    }
-
-    @Test
-    public void testCreateTwoWindingsTransformerInMixedTopology() throws Exception {
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/ModificationControllerTest.java:1306]: I cant find this test anymore.
the question is: where do we have to put it ?
I'm afraid it is using a third network ...

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationBusBreakerTest.java [hunk 27]
```diff
@@ -0,0 +1,436 @@
...
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+public class TwoWindingsTransformerCreationBusBreakerTest extends AbstractNetworkModificationTest {
+
+    @Override
+    protected Network createNetwork(UUID networkUuid) {
+        return NetworkCreation.createBusBreaker(networkUuid);
+    }
```
Comment (dbraquart) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationBusBreakerTest.java:41]: except this network creation, this test file looks like a big copy/paste of the NodeBreakerTest.
I dont know if it's relevant to test the BusBreaker view.
We need help to validate this.

And if it's good, we could use common code between the 2 test files?

**Review by dbraquart — CHANGES_REQUESTED** [review]
Front-end tests: I cannot create a 2WT modification with my local docker env, because your branch is too far behind the main branch. Could you rebase (and resolve conflicts), please ?

Code Review: still some points to clarify with BusBreaker tests. And 1 test is still missing.

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationNodeBreakerTest.java [hunk 28]
```diff
@@ -460,44 +460,6 @@ public void testCreateWithErrors() {
...
-                .andExpect(status().isOk()).andReturn();
-        resultAsString = mvcResult.getResponse().getContentAsString();
-        List<EquipmentModificationInfos> bsmlrTwoWindingsTransformer = mapper.readValue(resultAsString, new TypeReference<>() {
-        });
-        assertThat(bsmlrTwoWindingsTransformer.get(0), createMatcherEquipmentModificationInfos(ModificationType.TWO_WINDINGS_TRANSFORMER_CREATION, "id2wt1", Set.of("s1")));
-
-        assertNotNull(getNetwork().getTwoWindingsTransformer("id2wt1"));  // transformer was created
-        testNetworkModificationsCount(getGroupId(), 1);
-
         // Test create transformer on not yet existing variant VARIANT_NOT_EXISTING_ID :
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerCreationNodeBreakerTest.java:463]: Add a new line

**Review by SlimaneAmar — APPROVED** [review]

**Review by dbraquart — APPROVED** [review]
Front-end tests: OK
review: ok

**Pull request merged**